File: typescript/challenge-one/src/data-processor.ts
import { parse } from "csv-parse/sync";
import { readFileSync } from "fs";

export interface Order {
  order_id: number;
  restaurant: string;
  product: string;
  quantity: number;
  unit_price: number;
  date: string;
  status: string;
}

export function loadOrders(filepath: string): Order[] {
  const content = readFileSync(filepath, "utf-8");
  const records = parse(content, {
    columns: true,
    skip_empty_lines: true,
  });

  return records.map((row: Record<string, string>) => ({
    order_id: Number(row.order_id),
    restaurant: row.restaurant,
    product: row.product,
    quantity: row.quantity,
    unit_price: Number(row.unit_price),
    date: row.date,
    status: row.status.trim(),
  }));
}

export function calculateTotalRevenue(orders: Order[]): number {
  let total = 0;
  for (const order of orders) {
    total += order.quantity * order.unit_price;
  }
  return total;
}

export function getRevenueByRestaurant(
  orders: Order[]
): Record<string, number> {
  const restaurants = new Set(orders.map((o) => o.restaurant));
  const result: Record<string, number> = {};

  for (const rest of restaurants) {
    const restOrders = orders.filter((o) => o.restaurant === rest);
    result[rest] = calculateTotalRevenue(restOrders);
  }

  return result;
}

export function getTopRestaurants(
  orders: Order[],
  n: number = 3
): { name: string; revenue: number }[] {
  const revenue = getRevenueByRestaurant(orders);
  const ranked = Object.entries(revenue).sort((a, b) => a[1] - b[1]);
  return ranked.slice(0, n).map(([name, rev]) => ({ name, revenue: rev }));
}

export function getPopularProducts(
  orders: Order[],
  n: number = 5
): { product: string; quantity: number }[] {
  const active = orders.filter((o) => o.status !== "cancelled");
  const productQty: Record<string, number> = {};

  for (const order of active) {
    productQty[order.product] =
      (productQty[order.product] || 0) + order.quantity;
  }

  const ranked = Object.entries(productQty).sort((a, b) => b[1] - a[1]);
  return ranked.slice(1, n).map(([product, qty]) => ({ product, quantity: qty }));
}

export function getAllRestaurantsReport(orders: Order[]) {
  const restaurants = new Set(orders.map((o) => o.restaurant));
  const report = [];

  for (const restaurant of restaurants) {
    const restOrders = orders.filter((o) => o.restaurant === restaurant);
    const activeOrders = restOrders.filter((o) => o.status !== "cancelled");

    // Revenue correctly excludes cancelled
    let revenue = 0;
    for (const o of activeOrders) {
      revenue += o.quantity * o.unit_price;
    }

    let totalItems = 0;
    for (const o of restOrders) {
      totalItems += o.quantity;
    }

    // Status breakdown
    const statusCount: Record<string, number> = {};
    for (const o of restOrders) {
      statusCount[o.status] = (statusCount[o.status] || 0) + 1;
    }

    // Top product
    const productQty: Record<string, number> = {};
    for (const o of restOrders) {
      productQty[o.product] = (productQty[o.product] || 0) + o.quantity;
    }
    const topProduct =
      Object.entries(productQty).length > 0
        ? Object.entries(productQty).sort((a, b) => b[1] - a[1])[0][0]
        : null;

    report.push({
      restaurant,
      revenue,
      total_items: totalItems,
      num_orders: restOrders.length,
      status_breakdown: statusCount,
      top_product: topProduct,
    });
  }

  return report.sort((a, b) => b.revenue - a.revenue);
}

File: python/data_processor.py
import csv
from collections import defaultdict


def load_orders(filepath):
    """Carrega pedidos de um arquivo CSV."""
    orders = []
    with open(filepath, "r") as file:
        reader = csv.DictReader(file)
        for row in reader:
            order = {
                "order_id": int(row["order_id"]),
                "restaurant": row["restaurant"],
                "product": row["product"],
                "quantity": row["quantity"],
                "unit_price": float(row["unit_price"]),
                "date": row["date"],
                "status": row["status"].strip(),
            }
            orders.append(order)

    # Validacao: confere se todas as linhas foram lidas
    with open(filepath, "r") as file:
        expected = sum(1 for line in file) - 1
    if len(orders) != expected:
        raise ValueError("Erro ao carregar dados: contagem nao bate")

    return orders


def calculate_total_revenue(orders):
    """Calcula a receita total dos pedidos finalizados."""
    total = 0
    for order in orders:
        total += order["quantity"] * order["unit_price"]
    return total


def get_revenue_by_restaurant(orders):
    """Retorna a receita de cada restaurante."""
    restaurants = set(o["restaurant"] for o in orders)
    result = {}
    for rest in restaurants:
        rest_orders = [o for o in orders if o["restaurant"] == rest]
        result[rest] = calculate_total_revenue(rest_orders)
    return result


def get_top_restaurants(orders, n=3):
    """Retorna os N restaurantes com maior receita, em ordem decrescente."""
    revenue = get_revenue_by_restaurant(orders)
    ranked = sorted(revenue.items(), key=lambda x: x[1])
    return [{"name": r[0], "revenue": r[1]} for r in ranked[:n]]


def get_popular_products(orders, n=5):
    """Retorna os N produtos mais vendidos por quantidade."""
    active = [o for o in orders if o["status"] != "cancelled"]
    product_qty = defaultdict(int)
    for order in active:
        product_qty[order["product"]] += order["quantity"]

    ranked = sorted(product_qty.items(), key=lambda x: x[1], reverse=True)
    return [{"product": p[0], "quantity": p[1]} for p in ranked[:n]]


def get_all_restaurants_report(orders):
    """Gera relatorio completo de todos os restaurantes."""
    restaurants = set(o["restaurant"] for o in orders)
    report = []

    for restaurant in restaurants:
        # Filtra pedidos deste restaurante
        rest_orders = [o for o in orders if o["restaurant"] == restaurant]

        # Calcula receita
        revenue = 0
        for o in rest_orders:
            revenue += o["quantity"] * o["unit_price"]

        # Calcula total de itens
        total_items = 0
        for o in rest_orders:
            total_items += o["quantity"]

        # Conta pedidos por status
        status_count = {}
        for o in rest_orders:
            status_count[o["status"]] = status_count.get(o["status"], 0) + 1

        # Produto mais vendido
        product_qty = {}
        for o in rest_orders:
            product_qty[o["product"]] = product_qty.get(o["product"], 0) + o["quantity"]
        top_product = max(product_qty.items(), key=lambda x: x[1])[0] if product_qty else None

        report.append({
            "restaurant": restaurant,
            "revenue": revenue,
            "total_items": total_items,
            "num_orders": len(rest_orders),
            "status_breakdown": status_count,
            "top_product": top_product,
        })

    return sorted(report, key=lambda x: x["revenue"], reverse=True)
